src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
 
   return operands;
 }
+
+///////////
+abstract_value_pointert as_value(const abstract_object_pointert &obj)
+{
+  auto context_value =
+    std::dynamic_pointer_cast<const context_abstract_objectt>(obj);
+
+  return context_value
+           ? as_value(context_value->unwrap_context())
+           : std::dynamic_pointer_cast<const abstract_value_objectt>(obj);
+}
+
+bool is_value(const abstract_object_pointert &obj)
+{
+  return as_value(obj) != nullptr;
+}
+
+static auto inverse_operations =
+  std::map<irep_idt, irep_idt>{{ID_equal, ID_notequal},
+                               {ID_notequal, ID_equal},
+                               {ID_le, ID_gt},
+                               {ID_lt, ID_ge},
+                               {ID_ge, ID_lt},
+                               {ID_gt, ID_le}};
+
+exprt invert_expr(const exprt &expr)
+{
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
+
+  auto relation_expr = to_binary_relation_expr(expr);
+  auto inverse_op = inverse_operation->second;
+  return binary_relation_exprt(
+    relation_expr.lhs(), inverse_op, relation_expr.rhs());
+}
+
+exprt assume_not(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto not_expr = to_not_expr(expr);
+
+  auto inverse_expression = invert_expr(not_expr.op());
+  if(inverse_expression.is_not_nil())
+    return env.do_assume(inverse_expression, ns);
+
+  auto result = env.do_assume(not_expr.op(), ns);
+  if(result.is_boolean())
+    result =
+      result.is_true() ? static_cast<exprt>(false_exprt()) : true_exprt();
+  return result;
+}
+
+exprt assume_and(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto and_expr = to_and_expr(expr);
+  for(auto const &operand : and_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
+    if(result.is_nil() || result.is_false())
+      return result;
+  }
+  return true_exprt();
+}
+
+exprt assume_or(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  exprt or_result = false_exprt();
+  auto or_expr = to_or_expr(expr);
+  for(auto const &operand : or_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
```

Comment:
It is upsetting that this is not the dual of `assume_and` (and I fear I may not have communicated that very well) but this is not correct in the cases when `do_assume` can prune things (see comments below about pruning on ordering).  Consider the case of assuming `x == 1 || x == 2` as written this will (I think) result in the domain being bottom.

Instead I would have `assume_or(env, or_expr, ns, assuming_true)` call `assume_and(env, and_exprt(or_expr,operand()), ns, ! assume_true)`, i.e. you do the same thing as `assume_and` but inverting the each of the operands.

So rather than assuming `A || B`, you assume `!A && !B`, which means that if you prune the domain while handling `!A`, that is a good thing rather than conflicting with any of the pruning you do in `!B`.

Reply:
I've rewritten `assume_or` in terms of `assume_and`.